### DISCUSSION TITLE
Do we still need a maintenance release 3.3?

### DISCUSSION THREAD

**Discussion by FlorianWilhelm** [pyscaffold/pyscaffold#347] [General]
Right now there is an alpha version 3.3a1 release and some milestones which are 100% fullfilled. Does it make sense to release a version 3.3 or directly jump to 4.0?

**Comment by abravalheri** [comment]
I think we can release the 3.3 immediately (maybe just add the auto fix for the line endings with pre-commit?) since 4.0 seems like will take some time yet to be 100% ready

**Reply by FlorianWilhelm** [reply]
Alright, I'll work on that and prepare something it.

**Comment by FlorianWilhelm** [comment]
@abravalheri Was there an easy fix to the Windows unit tests? On the current master, I get on tests invoking `putup`:
```
stdout = 'running test\n******************************* DEPRECATED *******************************\npytest-runner is deprecated...91: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1124)\n'
```

**Reply by abravalheri** [reply]
That is an annoying error that always come back to bite us 😫 

Usually we have to update `pip` of the Python installation that is being invoked and also install `certifi` (apparently everytime the keyring for PyPI gets an update we have that error)... So if the tests use a virtual environment, that environment needs to be brought up to date...

But to be sincere I never understood this error, and sometimes I just mess around in the code until the error disappears :(
It is frustrating.

Which test file is raising this error? `test_install.py`?

**Reply by FlorianWilhelm** [reply]
Hm... strange. It's those tests but purely on Windows
```
FAILED tests/system/test_common.py::test_putup - subprocess.CalledProcessErro...
FAILED tests/system/test_common.py::test_differing_package_name - subprocess....
FAILED tests/system/test_common.py::test_extensions[travis-kwargs1-.travis.yml]
FAILED tests/system/test_common.py::test_extensions[django-kwargs3-manage.py]
FAILED tests/system/test_common.py::test_namespace - subprocess.CalledProcess...
FAILED tests/system/test_common.py::test_extensions[pre-commit-kwargs0-.pre-commit-config.yaml]
FAILED tests/system/test_common.py::test_extensions[gitlab-kwargs2-.gitlab-ci.yml]
FAILED tests/system/test_common.py::test_no_skeleton - subprocess.CalledProce...
FAILED tests/system/test_common.py::test_namespace_cookiecutter - subprocess....
```

**Reply by FlorianWilhelm** [reply]
Does that mean that tox is somehow caching the environments on Windows and we have to disable that cache and force it to reinstall `certifi` and `pip` every time?

**Reply by abravalheri** [reply]
I think it might be some optimisation of the pytest virtualenv plugin we are using in the master to create the virtualenvs.... I think last time I fixed this issue with this particular test file what I did is to stop relying on the plugin and started either avoiding the virtualenvs or using the venv extension in v4.

**Reply by abravalheri** [reply]
Hi Florian, I am curious, have you managed to solve the SSL error this time? If it just happens for Windows in v3.3 I would be happy to skip the tests, given it is clearly not a problem with our main code...

**Reply by FlorianWilhelm** [reply]
Yes, I solved it by backporting some of your code from the `4.0.x` branch, mainly `tests/systems/helpers.py`. It just seems as if under Windows tox somehow breaks out of the virtual environment in some cases and your code fixes this by explicitly calling the right interpreter and settings the env variables. It's unfortunate that we need so much boilerplate code to run unit tests but on the other hand, fixing these problems upstream would take so much longer.

**Reply by abravalheri** [reply]
That should be fine. I saw you also added a few env vars. Hopefully it will take time until we see another SSL error on windows again.

**Comment by abravalheri** [comment]
🥳  Thank you very much @FlorianWilhelm for working on v3.3 and the automatic release mechanism via github actions!!
